smart_open/asb.py

```diff
@@ -0,0 +1,400 @@
+
+_MIN_MIN_PART_SIZE = _REQUIRED_CHUNK_MULTIPLE = 4 * 1024**2
+"""Azure requires you to upload in multiples of 4MB, except for the last part."""
+
+_DEFAULT_MIN_PART_SIZE = 64 * 1024**2
+"""Default minimum part size for Azure Cloud Storage multipart uploads is 64MB"""
+
+DEFAULT_BUFFER_SIZE = 4 * 1024**2
+"""Default buffer size for working with Azure Storage Blob is 256MB
+https://docs.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs
+"""
+
+START = 0
+"""Seek to the absolute start of an Azure Storage Blob file"""
+
+CURRENT = 1
+"""Seek relative to the current positive of an Azure Storage Blob file"""
+
+END = 2
+"""Seek relative to the end of an Azure Storage Blob file"""
+
+_WHENCE_CHOICES = (START, CURRENT, END)
+
+_BINARY_NEWLINE = b'\n'
+
+def open(
+        bucket_id,
+        blob_id,
+        mode,
+        buffer_size=DEFAULT_BUFFER_SIZE,
+        client=None,  # type: azure.storage.blob.BlobServiceClient
+        ):
+    """Open an Azure Storage Blob blob for reading or writing.
+
+    Parameters
+    ----------
+    bucket_id: str
+        The name of the bucket this object resides in.
+    blob_id: str
+        The name of the blob within the bucket.
+    mode: str
+        The mode for opening the object.  Must be either "rb" or "wb".
+    buffer_size: int, optional
+        The buffer size to use when performing I/O. For reading only.
+    client: azure.storage.blob.BlobServiceClient, optional
+        The Azure Storage Blob client to use when working with azure-storage-blob.
+
+    """
+    if mode == _READ_BINARY:
+        return SeekableBufferedInputBase(
+            bucket_id,
+            blob_id,
+            buffer_size=buffer_size,
+            line_terminator=_BINARY_NEWLINE,
+            client=client,
+        )
+    elif mode == _WRITE_BINARY:
+        return BufferedOutputBase(
+            bucket_id,
+            blob_id,
+            client=client,
+        )
+    else:
+        raise NotImplementedError('GCS support for mode %r not implemented' % mode)
+
+
+class _SeekableRawReader(object):
+    """Read an Azure Storage Blob file."""
+
+    def __init__(self, asb_blob, size):
+        # type: (azure.storage.blob.BlobClient, int) -> None
+        self._blob = asb_blob
+        self._size = size
+        self._position = 0
+
+    def seek(self, position):
+        """Seek to the specified position (byte offset) in the Azure Storage Blob blob.
+
+        :param int position: The byte offset from the beginning of the blob.
+
+        Returns the position after seeking.
+        """
+        self._position = position
+        return self._position
+
+    def read(self, size=-1):
+        if self._position >= self._size:
+            return b''
+        binary = self._download_blob_chunk(size)
+        self._position += len(binary)
+        return binary
+
+    def _download_blob_chunk(self, size):
+        start = position = self._position
```

Comment:
Not sure if it works when we "rewind" the stream, aka call `seek` with a lower position than the current position.